design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+```
+
+This makes the output less awkward, but removes the type information in certain cases. In the following
+case it is not possible to derive whether `1` is a string or a number:
+
+```
+json_records("{\"a\": \"1\"}) // returns [["a", "1"]]
+```
+
+### json_concat
+
+```
+json_concat(json_string, json_string)
+```
+
+Given 2 strings, parse them as JSON values and return a string representing their concatenation.
+Concatenation rules are identical to PostgreSQL's [|| operator](https://www.postgresql.org/docs/14/functions-json.html):
+* If both strings deserialize into JSON objects, then return an object with a union of the input key,
+  taking values from the second object in the case of duplicates.
+* If at least one of the deserialized values is not an object, then convert non-array inputs to a
+  single-element array and return the result of array concatenation.
+* If at least one of the input strings is `NULL` or can't be deserialized as JSON, then return `NULL`.
+
+Akin to PostgreSQL's `||` operator, this function merges only top-level object keys or arrays.
+
+#### Examples
+
+```
+json_concat("{\"a\": 1}", "{\"b\": 2}") // returns "{\"a\": 1, \"b\": 2}"
+json_concat("{\"a\": {\"3\": 4}}", "{\"a\": {\"5\": 6}}") // returns "{\"a\": {\"5\": 6}}"
+json_concat("{\"a\": {\"5\": 6}}", "{\"a\": {\"3\": 4}}") // returns "{\"a\": {\"3\": 4}}"
+json_concat("{}", "{}") // returns "{}"
+json_concat("[1, 2]", "[3, 4]") // returns "[1,2,3,4]"
+json_concat("[1, [2]]", "[[[3]], [[[4]]]]") // returns "[ 1, [2], [[3]], [[[4]]] ]"
+json_concat("null", "null") // returns [null, null]
+json_concat("[1, 2]", "{\"a\": 1}") // returns "[1, 2, {\"a\": 1}]"
+json_concat("[1, 2]", "3") // returns "[1, 2, 3]"
+json_concat("[3]", "[1, 2]") // returns "[3, 1, 2]"
+json_concat("1", "2") // returns "[1, 2]"
+json_concat("[]", "[]") // returns []
+json_concat("abc", "[1]") // returns NULL
+json_concat(NULL, "[1]") // returns NULL
+```
+
+#### Rejected alternatives
+
+If one argument represents a valid JSON and the other one does not, then return the valid string
+as a result of concatenation.
+
+Given that concatenation of invalid JSON strings is an exceptional situation, I believe that we
+should return `NULL` if any input is not a JSON. This way, we avoid ambiguity if a user needs to trace
+back how they got certain results.
+
+If this function returns the valid record in the case when the other one is invalid, from the user
+perspective, there are 3 possibilities how they ended up with this result:
+* The second object could have been empty.
+* The second object could have had the same keys/values.
+* The second object is invalid JSON.
+
+By always returning `NULL` given an invalid JSON string we remove the third, the least obvious,
+possibility which reduces cognitive load from the user.
+
+### to_json_string
+
+```
+to_json_string(val)
+```
+
+Given any ksqlDB type returns the equivalent JSON string.
+
+#### Examples
```

Comment:
Can we add examples for all types? Missing: boolean, time, date, timestamp.

What about custom types? (Guess it's straightforward, but maybe worth to mention).

Might also be good to have a few example with nested complex types?

Reply:
Added in https://github.com/confluentinc/ksql/pull/8550/commits/4cd79269dcb217e5b43aa47ee420a85b720f72fb.